### DISCUSSION TITLE
how specify osx min version

### DISCUSSION THREAD

**Discussion by amicwall** [pyinstaller/pyinstaller#6316] [PyInstaller]
pyinstaller created my app on OSX 11.5 Big Sur
when trying to run on OSX 10.15 Catalina I get: You can't use this version of the application with this version of MacOS

Is there a min OSX version option somewhere, eg the spec file ?

**Comment by bwoodsend** [comment]
PyInstaller doesn't do any C compiling so a min OSX version parameter would be meaningless to it. It's the min OSX version of the binaries which PyInstaller has included which matters. Did you compile any packages from sdist or build Python from source (possibly via pyenv)? Both of those can lead to your having binaries which require your current OS's version instead of a more widely compatible one.

**Reply by rokm** [reply]
To expand the above explanation a bit; the min OSX version comes from two parts - one is PyInstaller's bootloader, the other are other binaries that PyInstaller collects (such as the Python shared library, but also binary extension modules both from base python and 3rd party python packages you use). So it boils down to how you installed python and how you installed PyInstaller.

PyInstaller comes with pre-compiled bootloaders that should target 10.13 and later. So if you install PyInstaller wheel (`pip install pyinstaller`), you should get that. If you compile bootloader yourself for some reason, it should still target 10.13 by default, unless you explicitly override `MACOSX_DEPLOYMENT_TARGET` environment variable. If you install PyInstaller via `brew`, then the bootloaders you are using were probably compiled with the "native" macOS version (i.e., the version you are running).

The other part is python shared library (and its extension modules, but those should be of same version). If you use python from python.org, the macOS min version should be low enough not to cause problems. If you built python from source (possibly via pyenv), then it depends on your build settings. If you installed python via `brew`, then you're out of luck because the min version is set to the macOS version you are running on.

Then there's also 3rd party packages with binary extensions (if you are using any), where similar limitations apply. Typically PyPI wheels should have lower min version, whereas the ones installed via `brew` will be targeting the running version.

So you should check the generated executable, the collected `Python` shared library, and any other collected `.so`/`.dylib` files to see which ones are (not) backwards compatible.

**Reply by amicwall** [reply]
thanks for your response. I installed python with brew, then: pip3 install pyinstaller 
So instead I should install python from python.org
Is there a command to check the min OSX version of the current python ?

**Reply by rokm** [reply]
>  Is there a command to check the min OSX version of the current python ?

You can use `otool` to list load commands of a binary (e.g., python interpreter itself, or the python shared library). For example, for python.org python 3.9 executable:

```bash
$ otool -l /usr/local/bin/python3.9 | grep LC_VERSION_MIN_MACOSX -A 3
      cmd LC_VERSION_MIN_MACOSX
  cmdsize 16
  version 10.9
      sdk 11.0
```

or for its shared library:

```bash
$ otool -l /Library/Frameworks/Python.framework/Versions/3.9/Python | grep LC_VERSION_MIN_MACOSX -A 3
      cmd LC_VERSION_MIN_MACOSX
  cmdsize 16
  version 10.9
      sdk 11.0
```

**Reply by amicwall** [reply]
great thanks, in my case with the brew python, there is no LC_VERSION_MIN_MACOSX
instead I had to type this:
otool -l /opt/homebrew/bin/python3.9 | grep LC_BUILD -A 4
      cmd LC_BUILD_VERSION
  cmdsize 32
 platform 1
    minos 11.0
      sdk 11.3

**Reply by rokm** [reply]
Ah yes, with newer targets, `LC_VERSION_MIN_MACOSX` was changed into `LC_BUILD_VERSION`.
